File: src/main/scala/com/blackwing/easyExploration/saveInventory/SaveInventory.scala
package com.blackwing.easyExploration.saveInventory

import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

import com.blackwing.easyExploration.config.Configuration
import com.blackwing.easyExploration.config.Configuration.InventoryOption
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.entity.player.{EntityPlayer, InventoryPlayer}
import net.minecraft.inventory.InventoryBasic
import net.minecraft.item.ItemStack
import net.minecraft.util.NonNullList
import org.apache.logging.log4j.Logger

import scala.collection.JavaConversions._
import scala.collection.mutable

object SaveInventory {

	private var logger: Logger = _

	def setLogger(logger: Logger): Unit = this.logger = logger

	private val config = Configuration.saveInventory
	private var inventories = mutable.Map[UUID, InventoryPlayer]()

	private[saveInventory] def put(player: EntityPlayer, inventory: InventoryPlayer) =
		inventories += player.getPersistentID -> inventory

	private[saveInventory] def get(player: EntityPlayer): InventoryPlayer =
		inventories.getOrElse(player.getPersistentID, new InventoryPlayer(player))

	private[saveInventory] def remove(player: EntityPlayer) =
		inventories -= player.getPersistentID

	private[saveInventory] def count(inv: InventoryPlayer): Int = {
		case null => 0
		case _ if inv.getSizeInventory <= 0 => 0
		case _ =>
			val itemStackCount = new AtomicInteger(0)
			for (stack <- inv.mainInventory) if (!stack.isEmpty) itemStackCount.incrementAndGet
			for (stack <- inv.armorInventory) if (!stack.isEmpty) itemStackCount.incrementAndGet
			for (stack <- inv.offHandInventory) if (!stack.isEmpty) itemStackCount.incrementAndGet
			itemStackCount.get
	}

	private[saveInventory] def count(inv: InventoryBasic): Int = {
		case null => 0
		case _ if inv.getSizeInventory <= 0 => 0
		case _ =>
			val itemStackCount = new AtomicInteger(0)
			for (slot <- Range(0, inv.getSizeInventory)) if (!inv.getStackInSlot(slot).isEmpty) itemStackCount.incrementAndGet
			itemStackCount.get
	}

	private def moveStacks(target: NonNullList[ItemStack], source: NonNullList[ItemStack]): Unit = {
		case (null, _) => logger.error("Target inventory missing.")
		case (_, null) => logger.error("Source inventory missing.")
		case _ if target.size != source.size => logger.error("Target inventory has wrong size.")
		case _ =>
			for (slot <- Range(0, target.size)) target.set(slot, source.get(slot))
			source.clear()
	}

	private def moveStacks(source: NonNullList[ItemStack], target: InventoryBasic): Unit = {
		case (null, _) => logger.warn("Source inventory missing.")
		case _ =>
			for (stack <- source) target.addItem(stack)
			source.clear()
	}

	private[saveInventory] def moveInventory(player: EntityPlayer) = {
		val inventory = new InventoryPlayer(player)
		val deathChest = new InventoryBasic(player.getName + "'s loot crate", true, inventory.armorInventory.size + inventory.offHandInventory.size + inventory.mainInventory.size)
		if (config.equipment eq InventoryOption.KEEP) {
			moveStacks(inventory.armorInventory, player.inventory.armorInventory)
			moveStacks(inventory.offHandInventory, player.inventory.offHandInventory)
		}
		if (config.loot eq InventoryOption.KEEP) moveStacks(inventory.mainInventory, player.inventory.mainInventory)
		if (config.equipment eq InventoryOption.SAVE) {
			moveStacks(player.inventory.armorInventory, deathChest)
			moveStacks(player.inventory.offHandInventory, deathChest)
		}
		if (config.loot eq InventoryOption.SAVE) moveStacks(player.inventory.mainInventory, deathChest)
		/*
		if (config.equipment == InventoryOption.DROP) {
			// happens automatically since we don't cancel the event and don't remove the items from the inventory
		}
		if (config.loot == InventoryOption.DROP) {
			// happens automatically since we don't cancel the event and don't remove the items from the inventory
		}
		*/
		inventories.put(player.getPersistentID, inventory)
		deathChest
	}

	private[saveInventory] def destroyVanishingCursedItems(inventory: InventoryPlayer): Unit = {
		for (slot <- Range(0, inventory.getSizeInventory)) {
			val itemstack = inventory.getStackInSlot(slot)
			if (!itemstack.isEmpty && EnchantmentHelper.hasVanishingCurse(itemstack)) inventory.removeStackFromSlot(slot)
		}
	}
}

File: src/main/java/com/blackwing/easyExploration/saveInventory/SaveInventory.java
package com.blackwing.easyExploration.saveInventory;

import com.blackwing.easyExploration.EasyExploration;
import com.blackwing.easyExploration.config.Configuration;
import com.blackwing.easyExploration.config.Configuration.InventoryOption;
import com.blackwing.easyExploration.inventory.InventoryPlayerEE;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

public class SaveInventory {

    /**
     * Singleton
     */
    private static final SaveInventory instance = new SaveInventory();

    /**
     * @return instance
     */
    public static SaveInventory instance() {
        return instance;
    }

    /**
     * Constructor
     */
    private SaveInventory() {
    }

    private static Configuration.SubCategorySaveInventory config = Configuration.saveInventory;
    private static Map<UUID, InventoryPlayerEE> inventories = new HashMap<UUID, InventoryPlayerEE>();

    public static void put(EntityPlayer player, InventoryPlayerEE inventory) {
        inventories.put(player.getPersistentID(), inventory);
    }

    public static InventoryPlayerEE get(EntityPlayer player) {
        return inventories.getOrDefault(player.getPersistentID(), new InventoryPlayerEE(player));
    }

    public static void remove(EntityPlayer player) {
        inventories.remove(player.getPersistentID());
    }

    static int count(InventoryPlayer inv) {
        if (inv == null) return 0;
        if (inv.getSizeInventory() <= 0) return 0;
        AtomicInteger itemStackCount = new AtomicInteger(0);
        for (ItemStack stack : inv.mainInventory) if (!stack.isEmpty()) itemStackCount.incrementAndGet();
        for (ItemStack stack : inv.armorInventory) if (!stack.isEmpty()) itemStackCount.incrementAndGet();
        for (ItemStack stack : inv.offHandInventory) if (!stack.isEmpty()) itemStackCount.incrementAndGet();
        return itemStackCount.get();
    }

    static int count(IInventory inv) {
        if (inv == null) return 0;
        if (inv.getSizeInventory() <= 0) return 0;
        AtomicInteger itemStackCount = new AtomicInteger(0);
        for (int i = 0; i < inv.getSizeInventory(); ++i)
            if (!inv.getStackInSlot(i).isEmpty()) itemStackCount.incrementAndGet();
        return itemStackCount.get();
    }

    private static void moveStacks(NonNullList<ItemStack> target, NonNullList<ItemStack> source) {
        if (target == null) {
            EasyExploration.logger.error("Target inventory missing.");
            return;
        }
        if (source == null) {
            EasyExploration.logger.error("Source inventory missing.");
            return;
        }
        if (target.size() != source.size()) {
            EasyExploration.logger.error("Target inventory has wrong size.");
            return;
        }
        for (int i = 0; i < target.size(); ++i) target.set(i, source.get(i));
        source.clear();
    }

    static InventoryPlayerEE keepInventory(EntityPlayer player) {
        InventoryPlayerEE inventory = new InventoryPlayerEE(player);

        if (config.equipment == InventoryOption.KEEP) {
            moveStacks(inventory.armorInventory, player.inventory.armorInventory);
            moveStacks(inventory.offHandInventory, player.inventory.offHandInventory);
        }
        if (config.loot == InventoryOption.KEEP) {
            moveStacks(inventory.mainInventory, player.inventory.mainInventory);
        }
        if (config.equipment == InventoryOption.SAVE) {
            inventory.inventoryDeathChest.moveStacks(player.inventory.armorInventory);
            inventory.inventoryDeathChest.moveStacks(player.inventory.offHandInventory);
        }
        if (config.loot == InventoryOption.SAVE) {
            inventory.inventoryDeathChest.moveStacks(player.inventory.mainInventory);
        }

        /*
        if (config.equipment == InventoryOption.DROP) {
            // happens automatically since we don't cancel the event and don't remove the items from the inventory
        }
        if (config.loot == InventoryOption.DROP) {
            // happens automatically since we don't cancel the event and don't remove the items from the inventory
        }
        */
        inventories.put(player.getPersistentID(), inventory);
        return inventory;
    }

    static void destroyVanishingCursedItems(InventoryPlayer inventory) {
        for (int i = 0; i < inventory.getSizeInventory(); ++i) {
            ItemStack itemstack = inventory.getStackInSlot(i);

            if (!itemstack.isEmpty() && EnchantmentHelper.hasVanishingCurse(itemstack)) {
                inventory.removeStackFromSlot(i);
            }
        }
    }


}
